zcm/julia/ZCM.jl

```diff
@@ -15,320 +35,253 @@ end
+    msgdata = unsafe_wrap(Vector{UInt8}, rbuf.data, rbuf.data_size)
+    msg = decode(opts.msgtype, msgdata)
+    opts.handler(channel, msg)
+    return nothing
+end
+
+function subscribe(zcm::Zcm, channel::AbstractString, options::T) where T <: SubscriptionOptions
+    # The C function is called with the receive buffer, channel name, and the 
+    # SubscriptionOptions struct
+    c_handler = cfunction(onresponse, Void, Tuple{Ref{Native.RecvBuf}, Cstring, Ref{T}})
+    uv_wrapper = ccall(("uv_zcm_msg_handler_create", "libzcmjulia"),
+                       Ptr{Native.UvSub}, 
+                       (Ptr{Void}, Ptr{Void}),
+                       c_handler, Ref(options))
+    uv_handler = cglobal(("uv_zcm_msg_handler_trigger", "libzcmjulia"))
+    csub = ccall(("zcm_subscribe", "libzcm"), Ptr{Native.Sub},
+                (Ptr{Native.Zcm}, Cstring, Ptr{Void}, Ptr{Native.UvSub}),
+                zcm, channel, uv_handler, uv_wrapper)
+    sub = Subscription(options, c_handler, uv_wrapper, uv_handler, csub)
+    push!(zcm.subscriptions, sub)
+    return sub
 end
-export subscribe;
 
-# RRR: might be a way to make this template a bit prettier
 # TODO: get this into docs:
 # Note to self: handler could be either a function or a functor, so long as it has
-#               handler(rbuf::RecvBuf, channel::String, msg::MsgType, usr) defined
-function subscribe(zcm::Zcm, channel::AbstractString, MsgType::DataType, handler, usr)
-    return subscribe(zcm, channel,
-                     function (rbuf::RecvBuf, channel::AbstractString, usr)
-                         msg = decode(MsgType, rbuf.data)
-                         handler(rbuf, channel, msg, usr);
-                     end, usr);
+#               handler(channel::String, msg::msgtype) defined
+function subscribe(zcm::Zcm, channel::AbstractString, msgtype::Type{<:AbstractZCMType}, handler::Function)
+    subscribe(zcm, channel, SubscriptionOptions(msgtype, handler))
 end
-export subscribe;
 
-function unsubscribe(zcm::Zcm, sub::Sub)
+function unsubscribe(zcm::Zcm, sub::Subscription)
     ret = ccall(("zcm_unsubscribe", "libzcm"), Cint,
-                 (Ptr{Native.Zcm}, Ptr{Native.Sub}), zcm.zcm, sub.sub);
+                 (Ptr{Native.Zcm}, Ptr{Native.Sub}), zcm, sub.native_sub)
     ccall(("uv_zcm_msg_handler_destroy", "libzcmjulia"), Void,
-          (Ptr{Native.UvSub},), sub.uv_wrapper);
-    return ret;
+          (Ptr{Native.UvSub},), sub.uv_wrapper)
+    deleteat!(zcm.subscriptions, findin(zcm.subscriptions, [sub]))
+    return ret
 end
-export unsubscribe;
 
-function publish(zcm::Zcm, channel::AbstractString, data::Array{UInt8}, datalen::UInt32)
+function publish(zcm::Zcm, channel::AbstractString, data::Vector{UInt8})
     return ccall(("zcm_publish", "libzcm"), Cint,
                  (Ptr{Native.Zcm}, Cstring, Ptr{Void}, UInt32),
-                 zcm.zcm, channel, data, datalen);
+                 zcm, convert(String, channel), data, length(data))
```

Comment:
Why the `convert()` to `String` here? Wont `ccall` convert the `channel` to a `CString` anyway?

Reply:
`ccall` can convert a `String` to a `Cstring`, but there are other kinds of AbstractString (most common is `SubString`, which is a shared view into another string). `ccall` can't convert those directly to `Cstring`, so we need to first convert them to plain Julia strings. `convert(String, channel)` is a no-op if `channel` is already a String, so there should be no performance penalty for this.

Reply:
perfect